com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js

```diff
@@ -1,247 +1,315 @@
 
-var Combine = {
-    _targets: [],
-    _targetIndex: 0,
-    // target: build target
+var GameArchiveLoader = {
+    // which files to load
+    _files: [],
+    _fileIndex: 0,
+    // file
     //  name: intended filepath of built object
     //  size: expected size of built object.
-    //  data: combined data
-    //  downloaded: total amount of data downloaded
+    //  data: combined pieces
+    //  downloaded: total bytes downloaded
     //  pieces: array of name, offset and data objects
     //  numExpectedFiles: total number of files expected in description
     //  lastRequestedPiece: index of last data file requested (strictly ascending)
-    //  totalLoadedPieces: counts the number of data files received
+    //  totalLoadedPieces: counts the number pieces received
 
     //MAX_CONCURRENT_XHR: 6,    // remove comment if throttling of XHR is desired.
 
     isCompleted: false,       // status of process
 
-    _onCombineCompleted: [],    // signature: name, data.
-    _onAllTargetsBuilt:[],      // signature: void
-    _onDownloadProgress: [],    // signature: downloaded, total
+    _onFileLoadedListeners: [],          // signature: name, data.
+    _onArchiveLoadedListeners:[],        // signature: void
+    _onDownloadProgressListeners: [],    // signature: downloaded, total
+    _onFileDownloadErrorListeners: [],   // signature: name
 
     _currentDownloadBytes: 0,
     _totalDownloadBytes: 0,
 
-    _retry_time: 0,             // pause before retry file loading after error
-    _max_retry_count: 0,        // how many attempts we do when trying to download a file.
-    _can_not_download_file_callback: undefined, //Function that is called if you can't download file after 'retry_count' attempts.
+    _retryTime: 0,             // pause before retry file loading after error
+    _maxRetryCount: 0,        // how many attempts we do when trying to download a file.
 
     _archiveLocationFilter: function(path) { return "split" + path; },
 
-    can_not_download_file: function(file) {
-        if (typeof Combine._can_not_download_file_callback === 'function') {
-            Combine._can_not_download_file_callback(file);
-        }
+    cleanUp: function() {
+        this._files =  [];
+        this._fileIndex = 0;
+        this.isCompleted = false;
+        this._onGameArchiveLoaderCompletedListeners = [];
+        this._onAllTargetsBuiltListeners = [];
+        this._onDownloadProgressListeners = [];
+        this._onFileDownloadErrorListeners = [];
+
+        this._currentDownloadBytes = 0;
+        this._totalDownloadBytes = 0;
     },
 
-    addProgressListener: function(callback) {
-        if (typeof callback !== 'function') {
-            throw "Invalid callback registration";
+    addListener: function(list, callback) {
+        if (typeof callback !== 'function') throw "Invalid callback registration";
+        list.push(callback);
+    },
+    notifyListeners: function(list, ...args) {
```

Comment:
Spread syntax (...) is an ES6 thing. https://caniuse.com/?search=spread%20operator
If you use it, old browsers can't parse `dmloader.js` and users [will get a blank page](https://user-images.githubusercontent.com/193258/93617606-dd5ae900-f9de-11ea-87cd-6b505bf42bee.png). 

Possible replacement:
```js
    notifyListeners: function(list) {
        for (i=0; i<list.length; ++i) {
            list[i].apply(list[i], Array.prototype.slice.call(arguments, 1));
        }
    },
```

Reply:
Oh, damn. Thanks.